// mypartner-route/lib/route-additional-configs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getProfileInfo = {
    tags: [
        'api', 'greeting', 'getProfileInfo'
    ],
    description: 'Get getProfileInfo.',
    validate: {
        params: {
        /*  msisdn: Joi
              .string()
              .required(),
          person_id:Joi.string()
              .required()*/
        }
    },
    plugins: {}
};
exports.orderTracking = {
    tags: [
        'api', 'greeting', 'orderTracking'
    ],
    description: 'Get recent activity data.',
    validate: {
        params: {
        /*  mobile_no: Joi.string().required(),
          provider_id:Joi.string().required(),
          action:Joi.string().required(),
          regis_no:Joi.string().required(),
          email:Joi.string().required()*/
        }
    },
    plugins: {}
};
exports.orderTrackingDetail = {
    tags: [
        'api', 'greeting', 'orderTrackingDetail'
    ],
    description: 'Get recent orderTrackingDetail data.',
    validate: {
        params: {
        /*    mobile_no: Joi.string().required(),
            provider_id:Joi.string().required(),
            action:Joi.string().required(),
            regis_no:Joi.string().required(),
            email:Joi.string().required()*/
        }
    },
    plugins: {}
};
exports.getItemPayment = {
    tags: [
        'api', 'greeting', 'getItemPayment'
    ],
    description: 'Get recent getItemPayment data.',
    validate: {
        params: {
        /*    mobile_no: Joi.string().required(),
            provider_id:Joi.string().required(),
            action:Joi.string().required(),
            regis_no:Joi.string().required(),
            email:Joi.string().required()*/
        }
    },
    plugins: {}
};
exports.updatePaymentCash = {
    tags: [
        'api', 'greeting', 'updatePaymentCash'
    ],
    description: 'Get recent updatePaymentCash data.',
    validate: {
        params: {
        /*    mobile_no: Joi.string().required(),
            provider_id:Joi.string().required(),
            action:Joi.string().required(),
            regis_no:Joi.string().required(),
            email:Joi.string().required()*/
        }
    },
    plugins: {}
};
exports.updatePaymentCredit = {
    tags: [
        'api', 'greeting', 'updatePaymentCredit'
    ],
    description: 'Get recent updatePaymentCredit data.',
    validate: {
        params: {
        /*    mobile_no: Joi.string().required(),
            provider_id:Joi.string().required(),
            action:Joi.string().required(),
            regis_no:Joi.string().required(),
            email:Joi.string().required()*/
        }
    },
    plugins: {}
};
exports.getBankAccount = {
    tags: [
        'api', 'greeting', 'getBankAccount'
    ],
    description: 'Get recent getBankAccount data.',
    validate: {
        params: {
        /*    mobile_no: Joi.string().required(),
            provider_id:Joi.string().required(),
            action:Joi.string().required(),
            regis_no:Joi.string().required(),
            email:Joi.string().required()*/
        }
    },
    plugins: {}
};
exports.auth = {
    tags: [
        'api', 'greeting', 'auth'
    ],
    description: 'Get recent auth data.',
    validate: {
        params: {
        /*    mobile_no: Joi.string().required(),
            provider_id:Joi.string().required(),
            action:Joi.string().required(),
            regis_no:Joi.string().required(),
            email:Joi.string().required()*/
        }
    },
    plugins: {}
};
exports.confirmOtp = {
    tags: [
        'api', 'greeting', 'confirmOtp'
    ],
    description: 'Get recent confirmOtp data.',
    validate: {
        params: {
        /*    mobile_no: Joi.string().required(),
            provider_id:Joi.string().required(),
            action:Joi.string().required(),
            regis_no:Joi.string().required(),
            email:Joi.string().required()*/
        }
    },
    plugins: {}
};
exports.login = {
    tags: [
        'api', 'greeting', 'login'
    ],
    description: 'Get recent login data.',
    validate: {
        params: {
        /*    mobile_no: Joi.string().required(),
            provider_id:Joi.string().required(),
            action:Joi.string().required(),
            regis_no:Joi.string().required(),
            email:Joi.string().required()*/
        }
    },
    plugins: {}
};
exports.menus = {
    tags: [
        'api', 'greeting', 'menus'
    ],
    description: 'Get recent menus data.',
    validate: {
        params: {
        /*    mobile_no: Joi.string().required(),
            provider_id:Joi.string().required(),
            action:Joi.string().required(),
            regis_no:Joi.string().required(),
            email:Joi.string().required()*/
        }
    },
    plugins: {}
};
exports.getCategoryName = {
    tags: [
        'api', 'greeting', 'getCategoryName'
    ],
    description: 'Get recent getCategoryName data.',
    validate: {
        params: {
        /*    mobile_no: Joi.string().required(),
            provider_id:Joi.string().required(),
            action:Joi.string().required(),
            regis_no:Joi.string().required(),
            email:Joi.string().required()*/
        }
    },
    plugins: {}
};
exports.getCategoryTrade = {
    tags: [
        'api', 'greeting', 'getCategoryTrade'
    ],
    description: 'Get recent getCategoryTrade data.',
    validate: {
        params: {
        /*    mobile_no: Joi.string().required(),
            provider_id:Joi.string().required(),
            action:Joi.string().required(),
            regis_no:Joi.string().required(),
            email:Joi.string().required()*/
        }
    },
    plugins: {}
};
exports.listProductCatalog = {
    tags: [
        'api', 'greeting', 'listProductCatalog'
    ],
    description: 'Get recent listProductCatalog data.',
    validate: {
        params: {
        /*    mobile_no: Joi.string().required(),
            provider_id:Joi.string().required(),
            action:Joi.string().required(),
            regis_no:Joi.string().required(),
            email:Joi.string().required()*/
        }
    },
    plugins: {}
};
exports.keepAlive = {
    tags: [
        'api', 'greeting', 'keepAlive'
    ],
    description: 'Get recent keepAlive data.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.validateSession = {
    tags: [
        'api', 'greeting', 'validateSession'
    ],
    description: 'Get recent validateSession data.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.getTradeDetail = {
    tags: [
        'api', 'greeting', 'getTradeDetail'
    ],
    description: 'Get recent getTradeDetail data.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.tradeFilter = {
    tags: [
        'api', 'greeting', 'tradeFilter'
    ],
    description: 'Get tradeFilter product.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.calTradeNormal = {
    tags: [
        'api', 'greeting', 'calTradeNormal'
    ],
    description: 'Get calTradeNormal data.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.splitOrder = {
    tags: [
        'api', 'greeting', 'splitOrder'
    ],
    description: 'Get splitOrder data.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.saveUser = {
    tags: [
        'api', 'greeting', 'saveUser'
    ],
    description: 'save recent user data.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.getUserToken = {
    tags: [
        'api', 'greeting', 'getUserToken'
    ],
    description: 'get user token data.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.createOrder = {
    tags: [
        'api', 'greeting', 'createOrder'
    ],
    description: 'create order data.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.chkPODuplicate = {
    tags: [
        'api', 'greeting', 'chkPODuplicate'
    ],
    description: 'create order data.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.cpcImage = {
    tags: [
        'api', 'greeting', 'cpcImage'
    ],
    description: 'create order data.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.announcement = {
    tags: [
        'api', 'greeting', 'announcement'
    ],
    description: 'Get announcement.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.announcementDealer = {
    tags: [
        'api', 'greeting', 'announcementDealer'
    ],
    description: 'Get announcementDealer.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.searchApi = {
    tags: [
        'api', 'greeting', 'searchApi'
    ],
    description: 'Post searchApi.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.getEncryptToken = {
    tags: [
        'api', 'greeting', 'getEncryptToken'
    ],
    description: 'Post getEncryptToken.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.getSearchApi = {
    tags: [
        'api', 'greeting', 'getSearchApi'
    ],
    description: 'Post getSearchApi.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.getUpdateVersion = {
    tags: [
        'api', 'greeting', 'getUpdateVersion'
    ],
    description: 'Post getUpdateVersion.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.queryAccBlock = {
    tags: [
        'api', 'greeting', 'queryAccBlock'
    ],
    description: 'Post queryAccBlock.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.releaseOrder = {
    tags: [
        'api', 'greeting', 'releaseOrder'
    ],
    description: 'Post releaseOrder.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.getArsOrder = {
    tags: [
        'api', 'greeting', 'getArsOrder'
    ],
    description: 'Post getArsOrder.',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.exportArs = {
    tags: [
        'api', 'greeting', 'exportArs'
    ],
    description: 'Post exportArs excel',
    validate: {
        params: {}
    },
    plugins: {}
};
exports.arslocation = {
    tags: [
        'api', 'greeting', 'arslocation'
    ],
    description: 'get ars location whitelist',
    validate: {
        params: {}
    },
    plugins: {}
};
