// mypartner-plugin/src/lib/repositories/check-po-repository.ts
import * as Bluebird from "bluebird";
import * as Moment from 'moment'
import {BaseRequestRepository} from "./base-request-repository"
import {CheckPODuplicateParams, IRepository, IRequest} from "./../interfaces"
require('tls').DEFAULT_CIPHERS= 'RSA'

export class CheckPORepository extends BaseRequestRepository implements IRepository {

    public registerMethod = (): void => {
        this.server.method({
            name: 'Regis.chkPODuplicate',
            method: this.chkPODuplicate,
            options: {
              callback: false
            }
        })
    }

    private chkPODuplicate = (request: any) => {
    
        let orderQueryOptions: CheckPODuplicateParams = {
            dealerCode:request.payload.dealerCode,
            listPoNo:request.payload.listPoNo,
        }
    
        // console.log('request', request);
        return this._queryCheckPODuplicate(request, orderQueryOptions)
    }

    private _queryCheckPODuplicate = (request: IRequest, poQueryOptions: CheckPODuplicateParams) => {
        let reqOptions = this._getReqOptionsPost(request)
        let now = Moment()
        reqOptions['certificate'] = 'tradingservice3.ais.co.th_2017.cer'
        reqOptions['uri'] = this.options['baseUrl'] + this.options.urlList['x-path-chkPODuplicate']
        reqOptions['headers']['Content-Type'] = 'application/json'
        reqOptions['body'] = {
            dealerCode:poQueryOptions.dealerCode,
            listPoNo:poQueryOptions.listPoNo
        }
        reqOptions['json'] = true
        const loggerName = 'tdm-service'
        return this.sendRequest(reqOptions, loggerName)
    }

}
