// mypartner-plugin/src/lib/repositories/order-tracking-detail-repository.ts
import * as Bluebird from "bluebird";
import * as Moment from 'moment'
import {BaseRequestRepository} from "./base-request-repository"
import {orderTrackingDetailParams, IRepository, IRequest} from "./../interfaces"
require('tls').DEFAULT_CIPHERS= 'RSA'

export class OrderTrackingDetailRepository extends BaseRequestRepository implements IRepository {

    public registerMethod = (): void => {
        this.server.method({
            name: 'Regis.orderTrackingDetail',
            method: this.orderTrackingDetail,
            options: {
              callback: false
            }
        })
    }

    private orderTrackingDetail = (request: any) => {
        let profileQueryOptions: orderTrackingDetailParams = {
            reqOrderNo: request.payload.reqOrderNo,
            sapOrderNo: request.payload.sapOrderNo
        }
        return this._queryOrderTrackingDetail(request, profileQueryOptions)
    }

    private _queryOrderTrackingDetail = (request: IRequest, profileQueryOptions: orderTrackingDetailParams) => {
        let reqOptions = this._getReqOptionsPost(request)
        let now = Moment()
      
        reqOptions['certificate'] = 'tradingservice3.ais.co.th_2017.cer'
        reqOptions['rejectUnauthorized'] = false
        reqOptions['uri'] = this.options['baseUrl'] + this.options.urlList['x-path-orderTrackingDetail']
        reqOptions['headers']['Content-Type'] = 'application/json'
        reqOptions['body'] = {
          reqOrderNo: profileQueryOptions.reqOrderNo,
          sapOrderNo: profileQueryOptions.sapOrderNo
        }
    
        reqOptions['json'] = true
        const loggerName = 'tdm-service'
        return this.sendRequest(reqOptions, loggerName)
      }
}
