// mypartner-plugin/src/lib/repositories/order-tracking-repository.ts
import * as Bluebird from "bluebird";
import * as Moment from 'moment'
import {BaseRequestRepository} from "./base-request-repository"
import {orderTrackingParams, IRepository, IRequest} from "./../interfaces"
require('tls').DEFAULT_CIPHERS= 'RSA'

export class OrderTrackingRepository extends BaseRequestRepository implements IRepository {

    public registerMethod = (): void => {
        this.server.method({
            name: 'Regis.orderTracking',
            method: this.orderTracking,
            options: {
              callback: false
            }
        })
    }

    private orderTracking = (request: any) => {
        let profileQueryOptions: orderTrackingParams = {
          reqDealerCode: request.payload.reqDealerCode,
          reqOrderNo: request.payload.reqOrderNo,
          sapPoNo: request.payload.sapPoNo,
          orderDateFrom: request.payload.orderDateFrom,
          orderDateTo: request.payload.orderDateTo,
          sapOrderNo: request.payload.sapOrderNo
        }
        return this._queryOrderTracking(request.url.query, profileQueryOptions)
    }
      
    private _queryOrderTracking = (request: IRequest, profileQueryOptions: orderTrackingParams) => {
        let reqOptions = this._getReqOptionsPost(request)
        let now = Moment()
        reqOptions['certificate'] = 'tradingservice3.ais.co.th_2017.cer'
        reqOptions['rejectUnauthorized'] = false
    

        reqOptions['uri'] = this.options['baseUrl'] + this.options.urlList['x-path-orderTracking']
        reqOptions['headers']['Content-Type'] = 'application/json'
    
        reqOptions['body'] = {
          locType: '',
          reqDealerCode: profileQueryOptions.reqDealerCode,
          matCode: '',
          reqOrderNo: profileQueryOptions.reqOrderNo,
          sapInvNo: '',
          sapPoNo: profileQueryOptions.sapPoNo,
          paymentMethod: '',
          createdBy: '',
          orderDateFrom: profileQueryOptions.orderDateFrom,
          orderDateTo: profileQueryOptions.orderDateTo,
          reqStatus: '',
          reqOrderGroup: '',
          sapOrderNo: profileQueryOptions.sapOrderNo
        }
    
        reqOptions['json'] = true
        const loggerName = 'tdm-service'
        return this.sendRequest(reqOptions, loggerName)
    }
    
}
