// mypartner-plugin/src/lib/repositories/product-catalog-repository.ts
import * as Bluebird from "bluebird";
import * as Moment from 'moment'
import {BaseRequestRepository} from "./base-request-repository"
import {ListProductCatalogParams, IRepository, IRequest} from "./../interfaces"
require('tls').DEFAULT_CIPHERS= 'RSA'

export class ProductCatalogRepository extends BaseRequestRepository implements IRepository {

    public registerMethod = (): void => {
        this.server.method({
            name: 'Regis.listProductCatalog',
            method: this.listProductCatalog,
            options: {
              callback: false
            }
        })
    }

    private listProductCatalog = (request: any) => {
        let profileQueryOptions: ListProductCatalogParams = {
          sessionId: request.payload.sessionId,
          page: request.payload.page,
          master: request.payload.master,
          paymentMethod: request.payload.paymentMethod,
          dealerCode: request.payload.dealerCode,
          customerGroup: request.payload.customerGroup,
          distChannel: request.payload.distChannel,
          locationCode: request.payload.locationCode,
          listPermission: request.payload.listPermission,
        }
        return this._queryListProductCatalog(request, profileQueryOptions)
    }

    private _queryListProductCatalog = (request: IRequest, profileQueryOptions: ListProductCatalogParams) => {
        let reqOptions = this._getReqOptionsPost(request)
        let now = Moment()
        // console.log('ConOtp =====>', this.options['authUrl'] + this.options.urlList['x-path-listProductCatalog'])
        reqOptions['certificate'] = 'tradingservice3.ais.co.th_2017.cer'
        reqOptions['uri'] = this.options['baseUrl'] + this.options.urlList['x-path-listProductCatalog']
        reqOptions['headers']['Content-Type'] = 'application/json'
        reqOptions['body'] = {
            sessionId: profileQueryOptions.sessionId,
            listPermission: profileQueryOptions.listPermission,
            page: profileQueryOptions.page,
            master: profileQueryOptions.master,
            paymentMethod: profileQueryOptions.paymentMethod,
            dealerCode: profileQueryOptions.dealerCode,
            customerGroup: profileQueryOptions.customerGroup,
            distChannel: profileQueryOptions.distChannel,
            locationCode: profileQueryOptions.locationCode
        }
        reqOptions['json'] = true
        const loggerName = 'tdm-service'
        return this.sendRequest(reqOptions, loggerName)
    }
    
    
}
