// mypartner-plugin/src/lib/repositories/update-payment-cash-repository.ts
import * as Bluebird from "bluebird";
import * as Moment from 'moment'
import {BaseRequestRepository} from "./base-request-repository"
import {updatePaymentCashParams, IRepository, IRequest} from "./../interfaces"
require('tls').DEFAULT_CIPHERS= 'RSA'

export class UpdatePaymentCashRepository extends BaseRequestRepository implements IRepository {

    public registerMethod = (): void => {
        this.server.method({
            name: 'Regis.updatePaymentCash',
            method: this.updatePaymentCash,
            options: {
              callback: false
            }
        })
    }

    private updatePaymentCash = (request: any) => {
        let profileQueryOptions: updatePaymentCashParams = {
          company: request.payload.company,
          dealerCode: request.payload.dealerCode,
          locationCode: request.payload.locationCode,
          totalOrder: request.payload.totalOrder,
          totalCn: request.payload.totalCn,
          totalDn: request.payload.totalDn,
          totalDiscount: request.payload.totalDiscount,
          totalInterest: request.payload.totalInterest,
          totalPayAmt: request.payload.totalPayAmt,
          paymentCat: request.payload.paymentCat,
          accountCode: request.payload.accountCode,
          userName: request.payload.userName,
          overdueWDS: request.payload.overdueWDS,
          overdueAWN_T: request.payload.overdueAWN_T,
          overdueAWN_O: request.payload.overdueAWN_O,
          productType: request.payload.productType,
          sapReceiptDtlList: request.payload.sapReceiptDtlList
        }
    
        // console.log('request cash ======>', profileQueryOptions);
        return this._queryPaymentCash(request, profileQueryOptions)
    }

    private _queryPaymentCash = (request: IRequest, profileQueryOptions: updatePaymentCashParams) => {
        let reqOptions = this._getReqOptionsPost(request)
        let now = Moment()
        reqOptions['certificate'] = 'tradingservice3.ais.co.th_2017.cer'
        reqOptions['rejectUnauthorized'] = false
    
        reqOptions['uri'] = this.options['baseUrl'] + this.options.urlList['x-path-updatePayment']
        reqOptions['headers']['Content-Type'] = 'application/json'
    
        reqOptions['body'] = {
          company: profileQueryOptions.company,
          dealerCode: profileQueryOptions.dealerCode,
          locationCode: profileQueryOptions.locationCode,
          totalOrder: profileQueryOptions.totalOrder,
          totalCn: profileQueryOptions.totalCn,
          totalDn: profileQueryOptions.totalDn,
          totalDiscount: profileQueryOptions.totalDiscount,
          totalInterest: profileQueryOptions.totalInterest,
          totalPayAmt: profileQueryOptions.totalPayAmt,
          paymentCat: profileQueryOptions.paymentCat,
          accountCode: profileQueryOptions.accountCode,
          userName: profileQueryOptions.userName,
          overdueWDS: profileQueryOptions.overdueWDS,
          overdueAWN_T: profileQueryOptions.overdueAWN_T,
          overdueAWN_O: profileQueryOptions.overdueAWN_O,
          productType: profileQueryOptions.productType,
          sapReceiptDtlList: profileQueryOptions.sapReceiptDtlList
        }
    
        reqOptions['json'] = true
    
        const loggerName = 'tdm-service'
        return this.sendRequest(reqOptions, loggerName)
    }
    
}
