// mypartner-plugin/lib/repositories/announcement-repository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Moment = require("moment");
const base_request_repository_1 = require("./base-request-repository");
require('tls').DEFAULT_CIPHERS = 'RSA';
class AnnouncementRepository extends base_request_repository_1.BaseRequestRepository {
    constructor() {
        super(...arguments);
        this.registerMethod = () => {
            this.server.method({
                name: 'Regis.announcement',
                method: this.announcement,
                options: {
                    callback: false
                }
            });
            this.server.method({
                name: 'Regis.announcementDealer',
                method: this.announcementDealer,
                options: {
                    callback: false
                }
            });
        };
        this.announcement = (request) => {
            let announceQueryOptions = {
                programCode: request.payload.programCode
            };
            // console.log('request', request);
            return this._queryAnnoucement(request, announceQueryOptions);
        };
        this._queryAnnoucement = (request, announceQueryOptions) => {
            let reqOptions = this._getReqOptionsGet(request);
            let now = Moment();
            reqOptions['certificate'] = 'tradingservice3.ais.co.th_2017.cer';
            reqOptions['rejectUnauthorized'] = false;
            reqOptions['uri'] =
                this.options['baseUrl'] +
                    this.options.urlList['x-path-getAnnouncement'] +
                    '?programCode=' +
                    announceQueryOptions.programCode;
            reqOptions['headers']['Content-Type'] = 'application/json';
            reqOptions['json'] = true;
            const loggerName = 'tdm-service';
            return this.sendRequest(reqOptions, loggerName);
        };
        this.announcementDealer = (request) => {
            let announceQueryOptions = {
                dealerCode: request.payload.dealerCode
            };
            // console.log('request', request);
            return this._queryAnnoucementDealer(request, announceQueryOptions);
        };
        this._queryAnnoucementDealer = (request, announceQueryOptions) => {
            let reqOptions = this._getReqOptionsGet(request);
            let now = Moment();
            reqOptions['certificate'] = 'tradingservice3.ais.co.th_2017.cer';
            reqOptions['rejectUnauthorized'] = false;
            reqOptions['uri'] =
                this.options['baseUrl'] +
                    this.options.urlList['x-path-getAnnouncementDealer'] +
                    '?dealerCode=' +
                    announceQueryOptions.dealerCode;
            reqOptions['headers']['Content-Type'] = 'application/json';
            reqOptions['json'] = true;
            const loggerName = 'tdm-service';
            return this.sendRequest(reqOptions, loggerName);
        };
    }
}
exports.AnnouncementRepository = AnnouncementRepository;
