// mypartner-plugin/lib/repositories/bank-account-repository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Moment = require("moment");
const base_request_repository_1 = require("./base-request-repository");
require('tls').DEFAULT_CIPHERS = 'RSA';
class BankAccountRepository extends base_request_repository_1.BaseRequestRepository {
    constructor() {
        super(...arguments);
        this.registerMethod = () => {
            this.server.method({
                name: 'Regis.getBankAccount',
                method: this.getBankAccount,
                options: {
                    callback: false
                }
            });
        };
        this.getBankAccount = (request) => {
            let profileQueryOptions = {
                userName: request.payload.userName
            };
            return this._queryBackAccount(request, profileQueryOptions);
        };
        this._queryBackAccount = (request, profileQueryOptions) => {
            let reqOptions = this._getReqOptionsGet(request);
            let now = Moment();
            reqOptions['certificate'] = 'tradingservice3.ais.co.th_2017.cer';
            reqOptions['rejectUnauthorized'] = false;
            reqOptions['uri'] =
                this.options['baseUrl'] +
                    this.options.urlList['x-path-getBankAccount'] +
                    '?userName=' +
                    profileQueryOptions.userName;
            reqOptions['headers']['Content-Type'] = 'application/json';
            reqOptions['json'] = true;
            const loggerName = 'tdm-service';
            return this.sendRequest(reqOptions, loggerName);
        };
    }
}
exports.BankAccountRepository = BankAccountRepository;
