// mypartner-plugin/lib/repositories/check-po-repository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Moment = require("moment");
const base_request_repository_1 = require("./base-request-repository");
require('tls').DEFAULT_CIPHERS = 'RSA';
class CheckPORepository extends base_request_repository_1.BaseRequestRepository {
    constructor() {
        super(...arguments);
        this.registerMethod = () => {
            this.server.method({
                name: 'Regis.chkPODuplicate',
                method: this.chkPODuplicate,
                options: {
                    callback: false
                }
            });
        };
        this.chkPODuplicate = (request) => {
            let orderQueryOptions = {
                dealerCode: request.payload.dealerCode,
                listPoNo: request.payload.listPoNo,
            };
            // console.log('request', request);
            return this._queryCheckPODuplicate(request, orderQueryOptions);
        };
        this._queryCheckPODuplicate = (request, poQueryOptions) => {
            let reqOptions = this._getReqOptionsPost(request);
            let now = Moment();
            reqOptions['certificate'] = 'tradingservice3.ais.co.th_2017.cer';
            reqOptions['uri'] = this.options['baseUrl'] + this.options.urlList['x-path-chkPODuplicate'];
            reqOptions['headers']['Content-Type'] = 'application/json';
            reqOptions['body'] = {
                dealerCode: poQueryOptions.dealerCode,
                listPoNo: poQueryOptions.listPoNo
            };
            reqOptions['json'] = true;
            const loggerName = 'tdm-service';
            return this.sendRequest(reqOptions, loggerName);
        };
    }
}
exports.CheckPORepository = CheckPORepository;
