// mypartner-plugin/lib/repositories/item-payment-repository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const base_request_repository_1 = require("./base-request-repository");
require('tls').DEFAULT_CIPHERS = 'RSA';
class ItemPaymentRepository extends base_request_repository_1.BaseRequestRepository {
    constructor() {
        super(...arguments);
        this.registerMethod = () => {
            this.server.method({
                name: 'Regis.getItemPayment',
                method: this.getItemPayment,
                options: {
                    callback: false
                }
            });
        };
        this.getItemPayment = (request) => {
            let profileQueryOptions = {
                dealerCode: request.payload.dealerCode,
                type: request.payload.type,
                userName: request.payload.userName
            };
            // console.log('request', request);
            return this._queryItemPayment(request, profileQueryOptions);
        };
        this._queryItemPayment = (request, profileQueryOptions) => {
            let reqOptions = this._getReqOptionsGet(request);
            reqOptions['certificate'] = 'tradingservice3.ais.co.th_2017.cer';
            reqOptions['rejectUnauthorized'] = false;
            reqOptions['uri'] =
                this.options['baseUrl'] +
                    this.options.urlList['x-path-getItemPayment'] +
                    '?dealerCode=' +
                    profileQueryOptions.dealerCode +
                    '&type=' +
                    profileQueryOptions.type +
                    '&userName=' +
                    profileQueryOptions.userName;
            reqOptions['headers']['Content-Type'] = 'application/json';
            reqOptions['json'] = true;
            const loggerName = 'tdm-service';
            return this.sendRequest(reqOptions, loggerName);
        };
    }
}
exports.ItemPaymentRepository = ItemPaymentRepository;
