// mypartner-plugin/lib/repositories/login-repository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Moment = require("moment");
const base_request_repository_1 = require("./base-request-repository");
require('tls').DEFAULT_CIPHERS = 'RSA';
class LoginRepository extends base_request_repository_1.BaseRequestRepository {
    constructor() {
        super(...arguments);
        this.registerMethod = () => {
            this.server.method({
                name: 'Regis.login',
                method: this.login,
                options: {
                    callback: false
                }
            });
        };
        this.login = (request) => {
            let profileQueryOptions = {
                tokenID: request.payload.tokenID
            };
            return this._queryLogin(request, profileQueryOptions);
        };
        this._queryLogin = (request, profileQueryOptions) => {
            let reqOptions = this._getReqOptionsPost(request);
            let now = Moment();
            reqOptions['certificate'] = 'tradingservice3.ais.co.th_2017.cer';
            reqOptions['uri'] = this.options['authUrl'] + this.options.urlList['x-path-login'];
            reqOptions['headers']['Content-Type'] = 'application/json';
            reqOptions['body'] = {
                tokenID: profileQueryOptions.tokenID
            };
            reqOptions['json'] = true;
            const loggerName = 'mychannel-service';
            return this.sendRequest(reqOptions, loggerName);
        };
        
    }
}
exports.LoginRepository = LoginRepository;
