// mypartner-plugin/lib/repositories/menus-repository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Moment = require("moment");
const base_request_repository_1 = require("./base-request-repository");
require('tls').DEFAULT_CIPHERS = 'RSA';
class MenusRepository extends base_request_repository_1.BaseRequestRepository {
    constructor() {
        super(...arguments);
        this.registerMethod = () => {
            this.server.method({
                name: 'Regis.menus',
                method: this.menus,
                options: {
                    callback: false
                }
            });
        };
        this.menus = (request) => {
            let profileQueryOptions = {
                xAuthorization: request.payload.xAuthorization
            };
            // console.log('request', request);
            return this._queryMenus(request, profileQueryOptions);
        };
        this._queryMenus = (request, profileQueryOptions) => {
            let reqOptions = this._getReqOptionsPost(request);
            let now = Moment();
            reqOptions['certificate'] = 'tradingservice3.ais.co.th_2017.cer';
            reqOptions['uri'] = this.options['authUrl'] + this.options.urlList['x-path-menus'];
            reqOptions['headers']['Content-Type'] = 'application/json';
            reqOptions['headers']['x-authorization'] = profileQueryOptions.xAuthorization;
            reqOptions['body'] = {};
            reqOptions['json'] = true;
            const loggerName = 'mychannel-service';
            return this.sendRequest(reqOptions, loggerName);
        };
    }
}
exports.MenusRepository = MenusRepository;
