// mypartner-plugin/lib/repositories/order-tracking-detail-repository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Moment = require("moment");
const base_request_repository_1 = require("./base-request-repository");
require('tls').DEFAULT_CIPHERS = 'RSA';
class OrderTrackingDetailRepository extends base_request_repository_1.BaseRequestRepository {
    constructor() {
        super(...arguments);
        this.registerMethod = () => {
            this.server.method({
                name: 'Regis.orderTrackingDetail',
                method: this.orderTrackingDetail,
                options: {
                    callback: false
                }
            });
        };
        this.orderTrackingDetail = (request) => {
            let profileQueryOptions = {
                reqOrderNo: request.payload.reqOrderNo,
                sapOrderNo: request.payload.sapOrderNo
            };
            return this._queryOrderTrackingDetail(request, profileQueryOptions);
        };
        this._queryOrderTrackingDetail = (request, profileQueryOptions) => {
            let reqOptions = this._getReqOptionsPost(request);
            let now = Moment();
            reqOptions['certificate'] = 'tradingservice3.ais.co.th_2017.cer';
            reqOptions['rejectUnauthorized'] = false;
            reqOptions['uri'] = this.options['baseUrl'] + this.options.urlList['x-path-orderTrackingDetail'];
            reqOptions['headers']['Content-Type'] = 'application/json';
            reqOptions['body'] = {
                reqOrderNo: profileQueryOptions.reqOrderNo,
                sapOrderNo: profileQueryOptions.sapOrderNo
            };
            reqOptions['json'] = true;
            const loggerName = 'tdm-service';
            return this.sendRequest(reqOptions, loggerName);
        };
    }
}
exports.OrderTrackingDetailRepository = OrderTrackingDetailRepository;
