// mypartner-plugin/lib/repositories/product-catalog-repository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Moment = require("moment");
const base_request_repository_1 = require("./base-request-repository");
require('tls').DEFAULT_CIPHERS = 'RSA';
class ProductCatalogRepository extends base_request_repository_1.BaseRequestRepository {
    constructor() {
        super(...arguments);
        this.registerMethod = () => {
            this.server.method({
                name: 'Regis.listProductCatalog',
                method: this.listProductCatalog,
                options: {
                    callback: false
                }
            });
        };
        this.listProductCatalog = (request) => {
            let profileQueryOptions = {
                sessionId: request.payload.sessionId,
                page: request.payload.page,
                master: request.payload.master,
                paymentMethod: request.payload.paymentMethod,
                dealerCode: request.payload.dealerCode,
                customerGroup: request.payload.customerGroup,
                distChannel: request.payload.distChannel,
                locationCode: request.payload.locationCode,
                listPermission: request.payload.listPermission,
            };
            return this._queryListProductCatalog(request, profileQueryOptions);
        };
        this._queryListProductCatalog = (request, profileQueryOptions) => {
            let reqOptions = this._getReqOptionsPost(request);
            let now = Moment();
            // console.log('ConOtp =====>', this.options['authUrl'] + this.options.urlList['x-path-listProductCatalog'])
            reqOptions['certificate'] = 'tradingservice3.ais.co.th_2017.cer';
            reqOptions['uri'] = this.options['baseUrl'] + this.options.urlList['x-path-listProductCatalog'];
            reqOptions['headers']['Content-Type'] = 'application/json';
            reqOptions['body'] = {
                sessionId: profileQueryOptions.sessionId,
                listPermission: profileQueryOptions.listPermission,
                page: profileQueryOptions.page,
                master: profileQueryOptions.master,
                paymentMethod: profileQueryOptions.paymentMethod,
                dealerCode: profileQueryOptions.dealerCode,
                customerGroup: profileQueryOptions.customerGroup,
                distChannel: profileQueryOptions.distChannel,
                locationCode: profileQueryOptions.locationCode
            };
            reqOptions['json'] = true;
            const loggerName = 'tdm-service';
            return this.sendRequest(reqOptions, loggerName);
        };
    }
}
exports.ProductCatalogRepository = ProductCatalogRepository;
