// mypartner-plugin/lib/repositories/profile-info-repository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Moment = require("moment");
const base_request_repository_1 = require("./base-request-repository");
require('tls').DEFAULT_CIPHERS = 'RSA';
class ProfileInfoRepository extends base_request_repository_1.BaseRequestRepository {
    constructor() {
        super(...arguments);
        this.registerMethod = () => {
            this.server.method({
                name: 'Regis.getProfileInfo',
                method: this.getProfileInfo,
                options: {
                    callback: false
                }
            });
        };
        this.getProfileInfo = (request) => {
            let profileQueryOptions = {
                locationCode: request.payload.locationCode
            };
            // console.log('request', request);
            return this._queryProfileInfo(request, profileQueryOptions);
        };
        this._queryProfileInfo = (request, profileQueryOptions) => {
            let reqOptions = this._getReqOptionsGet(request);
            let now = Moment();
            reqOptions['certificate'] = 'tradingservice3.ais.co.th_2017.cer';
            reqOptions['rejectUnauthorized'] = false;
            reqOptions['uri'] =
                this.options['baseUrl'] +
                    this.options.urlList['x-path-getProfileInfo'] +
                    '?locationCode=' +
                    profileQueryOptions.locationCode;
            reqOptions['headers']['Content-Type'] = 'application/json';
            reqOptions['json'] = true;
            const loggerName = 'tdm-service';
            return this.sendRequest(reqOptions, loggerName);
        };
    }
}
exports.ProfileInfoRepository = ProfileInfoRepository;
