// mypartner-plugin/lib/repositories/split-order-repository.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Moment = require("moment");
const base_request_repository_1 = require("./base-request-repository");
require('tls').DEFAULT_CIPHERS = 'RSA';
class SplitOrderRepository extends base_request_repository_1.BaseRequestRepository {
    constructor() {
        super(...arguments);
        this.registerMethod = () => {
            this.server.method({
                name: 'Regis.splitOrder',
                method: this.splitOrder,
                options: {
                    callback: false
                }
            });
        };
        this.splitOrder = (request) => {
            let splitOrderQueryOptions = {
                sessionId: request.payload.sessionId,
                userid: request.payload.userid,
                dealerCode: request.payload.dealerCode,
                locationCode: request.payload.locationCode,
                customerGroup: request.payload.customerGroup,
                distChannel: request.payload.distChannel,
                listCompany: request.payload.listCompany,
                shoppingCartList: request.payload.shoppingCartList,
            };
            // console.log('request', request);
            return this._querySplitOrder(request, splitOrderQueryOptions);
        };
        this._querySplitOrder = (request, splitOrderQueryOptions) => {
            let reqOptions = this._getReqOptionsPost(request);
            let now = Moment();
            // console.log('ConOtp =====>', this.options['authUrl'] + this.options.urlList['x-path-listProductCatalog'])
            reqOptions['certificate'] = 'tradingservice3.ais.co.th_2017.cer';
            reqOptions['uri'] = this.options['baseUrl'] + this.options.urlList['x-path-splitOrder'];
            reqOptions['headers']['Content-Type'] = 'application/json';
            reqOptions['body'] = {
                sessionId: splitOrderQueryOptions.sessionId,
                userid: splitOrderQueryOptions.userid,
                dealerCode: splitOrderQueryOptions.dealerCode,
                locationCode: splitOrderQueryOptions.locationCode,
                customerGroup: splitOrderQueryOptions.customerGroup,
                distChannel: splitOrderQueryOptions.distChannel,
                listCompany: splitOrderQueryOptions.listCompany,
                shoppingCartList: splitOrderQueryOptions.shoppingCartList,
            };
            reqOptions['json'] = true;
            const loggerName = 'tdm-service';
            return this.sendRequest(reqOptions, loggerName);
        };
    }
}
exports.SplitOrderRepository = SplitOrderRepository;
