// mypartner-route/src/lib/route-additional-configs.ts
import * as Joi from "joi";
import * as Hapi from "hapi";

export const getProfileInfo: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'getProfileInfo'
    ],
    description: 'Get getProfileInfo.',
    validate: {
        params: {
            /*  msisdn: Joi
                  .string()
                  .required(),
              person_id:Joi.string()
                  .required()*/
        }
    },
    plugins: {}
};

export const orderTracking: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'orderTracking'
    ],
    description: 'Get recent activity data.',
    validate: {
        params: {
            /*  mobile_no: Joi.string().required(),
              provider_id:Joi.string().required(),
              action:Joi.string().required(),
              regis_no:Joi.string().required(),
              email:Joi.string().required()*/

        }
    },
    plugins: {}
};

export const orderTrackingDetail: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'orderTrackingDetail'
    ],
    description: 'Get recent orderTrackingDetail data.',
    validate: {
        params: {
            /*    mobile_no: Joi.string().required(),
                provider_id:Joi.string().required(),
                action:Joi.string().required(),
                regis_no:Joi.string().required(),
                email:Joi.string().required()*/

        }
    },
    plugins: {}
};


export const getItemPayment: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'getItemPayment'
    ],
    description: 'Get recent getItemPayment data.',
    validate: {
        params: {
            /*    mobile_no: Joi.string().required(),
                provider_id:Joi.string().required(),
                action:Joi.string().required(),
                regis_no:Joi.string().required(),
                email:Joi.string().required()*/

        }
    },
    plugins: {}
};


export const updatePaymentCash: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'updatePaymentCash'
    ],
    description: 'Get recent updatePaymentCash data.',
    validate: {
        params: {
            /*    mobile_no: Joi.string().required(),
                provider_id:Joi.string().required(),
                action:Joi.string().required(),
                regis_no:Joi.string().required(),
                email:Joi.string().required()*/

        }
    },
    plugins: {}
};

export const updatePaymentCredit: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'updatePaymentCredit'
    ],
    description: 'Get recent updatePaymentCredit data.',
    validate: {
        params: {
            /*    mobile_no: Joi.string().required(),
                provider_id:Joi.string().required(),
                action:Joi.string().required(),
                regis_no:Joi.string().required(),
                email:Joi.string().required()*/

        }
    },
    plugins: {}
};

export const getBankAccount: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'getBankAccount'
    ],
    description: 'Get recent getBankAccount data.',
    validate: {
        params: {
            /*    mobile_no: Joi.string().required(),
                provider_id:Joi.string().required(),
                action:Joi.string().required(),
                regis_no:Joi.string().required(),
                email:Joi.string().required()*/

        }
    },
    plugins: {}
};

export const auth: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'auth'
    ],
    description: 'Get recent auth data.',
    validate: {
        params: {
            /*    mobile_no: Joi.string().required(),
                provider_id:Joi.string().required(),
                action:Joi.string().required(),
                regis_no:Joi.string().required(),
                email:Joi.string().required()*/

        }
    },
    plugins: {}
};

export const confirmOtp: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'confirmOtp'
    ],
    description: 'Get recent confirmOtp data.',
    validate: {
        params: {
            /*    mobile_no: Joi.string().required(),
                provider_id:Joi.string().required(),
                action:Joi.string().required(),
                regis_no:Joi.string().required(),
                email:Joi.string().required()*/

        }
    },
    plugins: {}
};

export const login: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'login'
    ],
    description: 'Get recent login data.',
    validate: {
        params: {
            /*    mobile_no: Joi.string().required(),
                provider_id:Joi.string().required(),
                action:Joi.string().required(),
                regis_no:Joi.string().required(),
                email:Joi.string().required()*/

        }
    },
    plugins: {}
};

export const menus: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'menus'
    ],
    description: 'Get recent menus data.',
    validate: {
        params: {
            /*    mobile_no: Joi.string().required(),
                provider_id:Joi.string().required(),
                action:Joi.string().required(),
                regis_no:Joi.string().required(),
                email:Joi.string().required()*/

        }
    },
    plugins: {}
};
export const getCategoryName: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'getCategoryName'
    ],
    description: 'Get recent getCategoryName data.',
    validate: {
        params: {
            /*    mobile_no: Joi.string().required(),
                provider_id:Joi.string().required(),
                action:Joi.string().required(),
                regis_no:Joi.string().required(),
                email:Joi.string().required()*/

        }
    },
    plugins: {}
};
export const getCategoryTrade: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'getCategoryTrade'
    ],
    description: 'Get recent getCategoryTrade data.',
    validate: {
        params: {
            /*    mobile_no: Joi.string().required(),
                provider_id:Joi.string().required(),
                action:Joi.string().required(),
                regis_no:Joi.string().required(),
                email:Joi.string().required()*/

        }
    },
    plugins: {}
};
export const listProductCatalog: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'listProductCatalog'
    ],
    description: 'Get recent listProductCatalog data.',
    validate: {
        params: {
            /*    mobile_no: Joi.string().required(),
                provider_id:Joi.string().required(),
                action:Joi.string().required(),
                regis_no:Joi.string().required(),
                email:Joi.string().required()*/
        }
    },
    plugins: {}
};

export const keepAlive: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'keepAlive'
    ],
    description: 'Get recent keepAlive data.',
    validate: {
        params: {
        }
    },
    plugins: {}
};


export const validateSession: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'validateSession'
    ],
    description: 'Get recent validateSession data.',
    validate: {
        params: {
        }
    },
    plugins: {}
};

export const getTradeDetail: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'getTradeDetail'
    ],
    description: 'Get recent getTradeDetail data.',
    validate: {
        params: {}
    },
    plugins: {}
};

export const tradeFilter: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'tradeFilter'
    ],
    description: 'Get tradeFilter product.',
    validate: {
        params: {}
    },
    plugins: {}
};
export const calTradeNormal: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'calTradeNormal'
    ],
    description: 'Get calTradeNormal data.',
    validate: {
        params: {}
    },
    plugins: {}
};
export const splitOrder: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'splitOrder'
    ],
    description: 'Get splitOrder data.',
    validate: {
        params: {}
    },
    plugins: {}
};
export const saveUser: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'saveUser'
    ],
    description: 'save recent user data.',
    validate: {
        params: {
        }
    },
    plugins: {}
};

export const getUserToken: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'getUserToken'
    ],
    description: 'get user token data.',
    validate: {
        params: {
        }
    },
    plugins: {}
};

export const createOrder: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'createOrder'
    ],
    description: 'create order data.',
    validate: {
        params: {
        }
    },
    plugins: {}
};

export const chkPODuplicate: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'chkPODuplicate'
    ],
    description: 'create order data.',
    validate: {
        params: {
        }
    },
    plugins: {}
};

export const cpcImage: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'cpcImage'
    ],
    description: 'create order data.',
    validate: {
        params: {
        }
    },
    plugins: {}
};

export const announcement: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'announcement'
    ],
    description: 'Get announcement.',
    validate: {
        params: {
        }
    },
    plugins: {}
};

export const announcementDealer: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'announcementDealer'
    ],
    description: 'Get announcementDealer.',
    validate: {
        params: {
        }
    },
    plugins: {}
};

export const searchApi: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'searchApi'
    ],
    description: 'Post searchApi.',
    validate: {
        params: {
        }
    },
    plugins: {}
};

export const getEncryptToken: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'getEncryptToken'
    ],
    description: 'Post getEncryptToken.',
    validate: {
        params: {
        }
    },
    plugins: {}
};

export const getSearchApi: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'getSearchApi'
    ],
    description: 'Post getSearchApi.',
    validate: {
        params: {
        }
    },
    plugins: {}
};

export const getUpdateVersion: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'getUpdateVersion'
    ],
    description: 'Post getUpdateVersion.',
    validate: {
        params: {
        }
    },
    plugins: {}
};


export const queryAccBlock: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'queryAccBlock'
    ],
    description: 'Post queryAccBlock.',
    validate: {
        params: {
        }
    },
    plugins: {}
};


export const releaseOrder: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'releaseOrder'
    ],
    description: 'Post releaseOrder.',
    validate: {
        params: {
        }
    },
    plugins: {}
};

export const getArsOrder: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'getArsOrder'
    ],
    description: 'Post getArsOrder.',
    validate: {
        params: {
        }
    },
    plugins: {}
};


export const exportArs: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'exportArs'
    ],
    description: 'Post exportArs excel',
    validate: {
        params: {
        }
    },
    plugins: {}
};



export const arslocation: Hapi.IRouteAdditionalConfigurationOptions = {
    tags: [
        'api', 'greeting', 'arslocation'
    ],
    description: 'get ars location whitelist',
    validate: {
        params: {}
    },
    plugins: {}
};
